### PR TITLE
[#180] add multipart support

### PR DESCRIPTION
**Pull request by RemboL**

### CODE REVIEW

File: accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy [hunk 1]
```diff
@@ -122,6 +122,17 @@ abstract class SpockMethodBodyBuilder {
 		return trimRepeatedQuotes(json)
 	}
 
+	protected Map<String, Object> getMultipartParameters() {
+		return (Map<String, Object>)request.multipart.serverValue
+	}
+
+	protected String getMultipartParameterLine(Map.Entry<String, Object> parameter) {
+		if (parameter.key == 'file') {
```
Comment (RemboL) [inline accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy]: @wszwech I don't think we should fixate on parameter name (in multipart you can send multiple streamed params with different key), please change this to something like parameter.value instanceof NamedProperty

**Comment by mariuszs** [comment]
:+1:

File: accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy [hunk 2]
```diff
@@ -122,6 +122,17 @@ abstract class SpockMethodBodyBuilder {
 		return trimRepeatedQuotes(json)
 	}
 
+	protected Map<String, Object> getMultipartParameters() {
+		return (Map<String, Object>)request.multipart.serverValue
```
Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy]: can `multipart` be null? IF so you can get a NPE

File: accurest-core/src/main/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBodyBuilder.groovy [hunk 3]
```diff
@@ -28,6 +29,9 @@ class MockMvcSpockMethodBodyBuilder extends SpockMethodBodyBuilder {
 		if (request.body) {
 			bb.addLine(".body('''$bodyAsString''')")
 		}
+		if (request.multipart) {
+			multipartParameters.each { entry -> bb.addLine(getMultipartParameterLine(entry)) }
```
Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBodyBuilder.groovy]: Can you add type to `Entry`?

File: accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Common.groovy [hunk 4]
```diff
@@ -48,6 +48,14 @@ class Common {
 		return property
 	}
 
+	NamedProperty named(DslProperty name, DslProperty value){
+		return new NamedProperty(name, value)
+	}
+
+	NamedProperty named(Map<String, DslProperty> namedMap){
+		return new NamedProperty(namedMap.get('name'), namedMap.get('content'))
```
Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Common.groovy]: Maybe extract to some descriptive constants?

File: accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Multipart.groovy [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+@EqualsAndHashCode(includeFields = true)
+@CompileStatic
+class Multipart extends DslProperty {
+
+    Multipart(Map<String, DslProperty> multipart) {
+        super(extractValue(multipart, { DslProperty p -> p.clientValue}), extractValue(multipart, {DslProperty p -> p.serverValue}))
+    }
+
+    private static Map<String, Object> extractValue(Map<String, DslProperty> multipart, Closure valueProvider) {
+        multipart.collectEntries { Map.Entry<String, DslProperty> entry ->
```
Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Multipart.groovy]: `return` is missing

File: accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Request.groovy [hunk 6]
```diff
@@ -100,6 +102,26 @@ class Request extends Common {
...
+
+	void multipart(DslProperty dslProperty) {
+		this.multipart = new Multipart(dslProperty)
+	}
+
+	void multipart(Object multipartAsValue) {
+		this.multipart = new Multipart(multipartAsValue)
+	}
+
+	Multipart getMultipart() {
```
Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Request.groovy]: you don't need this method casue

`Multipart multipart` generates a getter and setter for you

File: accurest-core/src/test/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBuilderSpec.groovy [hunk 7]
```diff
@@ -836,4 +838,65 @@ World.''')
...
+		given:
+			GroovyDsl contractDsl = GroovyDsl.make {
+				request {
+					method "PUT"
+					url "/multipart"
+					headers {
+						header('content-type', 'multipart/form-data;boundary=AaB03x')
+					}
+					multipart(
+							formParameter: value(client(regex('.+')), server('"formParameterValue"')),
```
Comment (marcingrzejszczak) [inline accurest-core/src/test/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBuilderSpec.groovy:852]: I can't find it but are you checking if the server value is matching the regex in on the client side?

File: accurest-core/src/test/groovy/io/codearte/accurest/dsl/WireMockGroovyDslSpec.groovy [hunk 8]
```diff
@@ -1398,4 +1398,50 @@ class WireMockGroovyDslSpec extends Specification implements WireMockStubVerifie
...
+			String wireMockStub = new WireMockStubStrategy(contractDsl).toWireMockClientStub()
+		then:
+			println wireMockStub
+			AssertionUtil.assertThatJsonsAreEqual(('''
+		{
+		  "request" : {
+			"url" : "/multipart",
+			"method" : "PUT",
+			"bodyPatterns" : [ {
+				"matches" : ".*--(.*)\\r\\nContent-Disposition: form-data; name=\\"formParameter\\"\\r\\n(Content-Type: .*\\r\\n)?(Content-Length: \\\\d+\\r\\n)?\\r\\n\\".+\\"\\r\\n--\\\\1.*"
```
Comment (marcingrzejszczak) [inline accurest-core/src/test/groovy/io/codearte/accurest/dsl/WireMockGroovyDslSpec.groovy:1432]: I believe you that it should look like this :D

**Comment by marcingrzejszczak** [comment]
Other than those small fixes a definite +1 ! Congrats @RemboL

File: accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy (hunk 2 above)
Comment (RemboL) [inline accurest-core/src/main/groovy/io/codearte/accurest/builder/SpockMethodBodyBuilder.groovy]: so far the only usage is conditioned with if (request.multipart), but I guess some "?."s won't hurt

File: accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Multipart.groovy (hunk 5 above)
Comment (RemboL) [inline accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Multipart.groovy]: Technically, it's not - http://groovy-lang.org/semantics.html#_optional_return_keyword , but I guess it'll make it clearer

Comment (marcingrzejszczak) [inline accurest-core/src/main/groovy/io/codearte/accurest/dsl/internal/Multipart.groovy]: yes, it's optional but we prefer to have it explicit (only in small closures we can omit this)

File: accurest-core/src/test/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBuilderSpec.groovy (hunk 7 above)
Comment (RemboL) [inline accurest-core/src/test/groovy/io/codearte/accurest/builder/MockMvcSpockMethodBuilderSpec.groovy:852]: not explicitly, as far as I see this assertion is part of common logic of value method.

**Pull request merged**